app/controllers/users_controller.rb

```diff
@@ -75,23 +74,71 @@ def photo
       type: @user.photo_content_type, disposition: :inline
   end
 
+  # GET users/1/change_email
+  def change_email
+  end
+
+  # GET users/1/change_password
+  def change_password
+  end
+
+  # PATCH users/1/update_email
+  def update_email
+    respond_to do |format|
+      if !@user.authenticate!(params[:user][:password], :password)
+        error_msg = t('.incorrect_password')
+        flash.now[:alert] = error_msg
+        format.html { render :change_email }
+        format.json { render json: {error: error_msg}, status: :unprocessable_entity }
+      elsif @user.update(email: params[:user][:email])
+        format.html { redirect_to @user, notice: t('.notice') }
+        format.json { render :show, status: :ok, location: @user }
+      else
+        error_msg = t('.invalid_email')
+        flash.now[:alert] = error_msg
+        format.html { render :change_email }
+        format.json { render json: {error: error_msg}, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # PATCH users/1/update_password
+  def update_password
+    respond_to do |format|
+      if !@user.authenticate!(params[:user][:current_password], :current_password)
```

Comment:
We're getting the following message when accessing this route: `Unpermitted parameter: :current_password`. Shoud we also permit this field for password update?

Reply:
Yep. I forget to "permit" key in password_update_params method.